### DISCUSSION TITLE
WeAct black pill multiple i2c ports

### DISCUSSION THREAD

**Discussion by Dave-Poissant** [zephyrproject-rtos/zephyr#43782] [Q&A] [answered]
I'm working with the WeAct black pill V2.0 (stm32f411ce) for a project. 

I've installed the zephyr RTOS blinky project on VSCode to start working with the board and i've managed to establish an I2C communication through the i2c1 node (pins B8(SCL) and B9(SDA)) which is the default i2c mapped bus in a zephyr project.

My problem now is that I'm trying to enable the 2 other i2c ports on the board but doesn't seem to be able to get it to work.

**What've done:**
- Managed to change the generated zephyr.dts file with an .overlay file:
/ {
    aliases {
            i2c1node = &i2c1;
            i2c2node = &i2c2;
    };
};

&{/soc/i2c@40005800/} {
    pinctrl-0 = <&i2c1_scl_pb6 &i2c1_sda_pb7>;
    status = "okay";
};

- Enabled config_I2C in proj.conf:
CONFIG_LOG=y
CONFIG_I2C=y

- My master code (working with default i2c pins)
- My arduino atmega2560 as the Slave

**The problem**
Code is compiling, but any maccros where im referencing either directly "i2c2" or my alias doesn't seem to return any usable device pointer.

**My question:**
First, is it possible to change the default i2c ports through overlays or other kind of configuration file? Then, is it possible to have more than 1 i2c port handler with zephyr? And finally, if those are possible what am I missing here to get it to work?

If needed, I could post my code in another comment.

**Comment by erwango** [comment]
Let's start from a known working point

In blackpill_f411ce, i2c1 node is enabled in the following way
https://github.com/zephyrproject-rtos/zephyr/blob/1add69b5c39377a19899a715f3db1c01c5b2f2c8/boards/arm/blackpill_f411ce/blackpill_f411ce.dts#L96-L102

If you need to enable to enable i2c2, you can add the following in a .overlay:
```
 &i2c2 { 
 	pinctrl-0 = <&i2c2_scl_pyz &i2c1_sda_pyz>; => To be configured depending on your board/needs
 	pinctrl-names = "default"; 
 	status = "okay"; 
 	clock-frequency = <I2C_BITRATE_FAST>; 
 }; 
```

In your code:
```
#define I2C2_DEV_NAME	DT_LABEL(DT_ALIAS(i2c2node))
const struct device *i2cé_dev = device_get_binding(I2C2_DEV_NAME);
```

**Reply by Dave-Poissant** [reply]
Sorry for the delay!

I've tryed what you proposed, seems like
`&{/soc/i2c@40005800/} {
pinctrl-0 = <&i2c1_scl_pb6 &i2c1_sda_pb7>;
status = "okay";
};`
is not the same as what you proposed, pretty sure its because i was getting the 
`&{/soc/i2c@40005800/}`
from the generated zephyr.dts file and not from the <BOARD>.dts

I now confirm that the Maccros seem to find the i2c2node (which wasn't the case before) **but I'm still having trouble with the i2c_transfer not returning 0.**

I'll show my atmega2560 slave code and my blackpill's master code. The blackpill's i2c transfer functions were strongly inspired by this fujitsu_fram i2c example https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/i2c_fujitsu_fram/src/main.c

blackpill (don't mind the LED related code, it's my debug tool since I don't have a prob yet):
<img width="522" alt="image" src="https://user-images.githubusercontent.com/43071732/158867684-b3b67d1d-bcbf-4605-900e-517565965d6b.png">
<img width="513" alt="image" src="https://user-images.githubusercontent.com/43071732/158867791-4d253eb1-c222-469e-bfe4-9abc822fe078.png">
<img width="620" alt="image" src="https://user-images.githubusercontent.com/43071732/158868032-a74bc586-a276-4834-af2b-7adee74bcfa0.png">
<img width="608" alt="image" src="https://user-images.githubusercontent.com/43071732/158868097-49ac1b5f-5e64-4279-8c2f-a0d2388b9ee1.png">
<img width="574" alt="image" src="https://user-images.githubusercontent.com/43071732/158868161-67a7c3f8-7ab6-45ec-a56d-e7ff21558ea3.png">


atmega2560 (slave):
<img width="423" alt="image" src="https://user-images.githubusercontent.com/43071732/158868250-3d6e97c7-a424-4278-900a-8db5debc7b0e.png">
<img width="373" alt="image" src="https://user-images.githubusercontent.com/43071732/158868287-50b81c41-5b27-4809-9c18-b18998c87f57.png">

**Reply by erwango** [reply]
> but I'm still having trouble with the i2c_transfer not returning 0.

I'd suggest checking the returned code, and start from there (it could be caused by anything (wires, code, ...)).

**Reply by Dave-Poissant** [reply]
Return value is "-5". According to official documentation https://docs.zephyrproject.org/latest/reference/peripherals/i2c.html, i2c_transfer returns:

- 0 – If successful.
-EIO – General input / output error.

Not sure what "EIO" means

**Comment by Dave-Poissant** [comment, marked as answer]
Figured it out.

Pins PB6-PB7 are still the i2c1 bus. I was interacting with it as an "i2c2" bus and it seems it didn't enjoy it. Changed the mux-pins to scl_pb10 and sda_pb3.

I find it a bit weird tho that the the bus didn't work just because its name was not the good one.

Thank you for your time @erwango
